File: Day050_2025/Q050_Leetcode.py
class Solution:
    def getHappyString(self, n: int, k: int) -> str:
        total = 3 * (1 << (n - 1))  # Total happy strings
        if k > total:
            return ""

        result = ['a'] * n
        next_smallest = {'a': 'b', 'b': 'a', 'c': 'a'}
        next_greatest = {'a': 'c', 'b': 'c', 'c': 'b'}

        startA, startB, startC = 1, 1 + (1 << (n - 1)), 1 + 2 * (1 << (n - 1))

        if k < startB:
            result[0] = 'a'
            k -= startA
        elif k < startC:
            result[0] = 'b'
            k -= startB
        else:
            result[0] = 'c'
            k -= startC

        for char_index in range(1, n):
            midpoint = (1 << (n - char_index - 1))
            if k < midpoint:
                result[char_index] = next_smallest[result[char_index - 1]]
            else:
                result[char_index] = next_greatest[result[char_index - 1]]
                k -= midpoint

        return "".join(result)

File: Day050_2025/Q050_Leetcode.java
import java.util.*;

class Q050_Leetcode {
    public String getHappyString(int n, int k) {
        int total = 3 * (1 << (n - 1)); // Total happy strings
        if (k > total) return "";

        StringBuilder result = new StringBuilder();
        Map<Character, Character> nextSmallest = Map.of('a', 'b', 'b', 'a', 'c', 'a');
        Map<Character, Character> nextGreatest = Map.of('a', 'c', 'b', 'c', 'c', 'b');

        int startA = 1;
        int startB = startA + (1 << (n - 1));
        int startC = startB + (1 << (n - 1));

        if (k < startB) {
            result.append('a');
            k -= startA;
        } else if (k < startC) {
            result.append('b');
            k -= startB;
        } else {
            result.append('c');
            k -= startC;
        }

        for (int charIndex = 1; charIndex < n; charIndex++) {
            int midpoint = (1 << (n - charIndex - 1));
            if (k < midpoint) {
                result.append(nextSmallest.get(result.charAt(charIndex - 1)));
            } else {
                result.append(nextGreatest.get(result.charAt(charIndex - 1)));
                k -= midpoint;
            }
        }

        return result.toString();
    }
}
